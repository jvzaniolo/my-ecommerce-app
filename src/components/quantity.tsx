import { MdAdd, MdRemove } from 'react-icons/md'
import {
  HStack,
  Icon,
  IconButton,
  Input,
  useNumberInput,
} from '@chakra-ui/react'

type QuantityProps = {
  max: number
  value: number
  onChange: (value: number) => void
}

const Quantity = ({ max, value, onChange }: QuantityProps) => {
  const { getInputProps, getIncrementButtonProps, getDecrementButtonProps } =
    useNumberInput({
      min: 1,
      max,
      value,
      onChange: (_, number) => {
        onChange(number)
      },
    })

  const inc = getIncrementButtonProps()
  const dec = getDecrementButtonProps()
  const input = getInputProps()

  return (
    <HStack>
      <IconButton
        icon={<Icon as={MdRemove} />}
        aria-label="Decrease"
        data-testid="decrease-quantity"
        {...dec}
      />
      <Input type="number" name="quantity" {...input} />
      <IconButton
        icon={<Icon as={MdAdd} />}
        aria-label="Increase"
        data-testid="increase-quantity"
        {...inc}
      />
    </HStack>
  )
}

export default Quantity
