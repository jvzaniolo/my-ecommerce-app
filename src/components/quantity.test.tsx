import { render } from '@testing-library/react'
import { Quantity } from './quantity'

describe('Quantity', () => {
  it('should render the quantity with initial props', () => {
    const { container } = render(
      <Quantity value={1} onChange={() => {}} max={100} />
    )

    expect(container).toMatchSnapshot()
  })

  it('should increase the quantity when the increase button is clicked', () => {
    const onChange = jest.fn()
    const { getByTestId } = render(
      <Quantity value={1} onChange={onChange} max={100} />
    )

    const increaseButton = getByTestId('increase-quantity')
    increaseButton.click()

    expect(onChange).toHaveBeenCalledWith(2)
  })

  it('should decrease the quantity when the decrease button is clicked', () => {
    const onChange = jest.fn()
    const { getByTestId } = render(
      <Quantity value={2} onChange={onChange} max={100} />
    )

    const decreaseButton = getByTestId('decrease-quantity')
    decreaseButton.click()

    expect(onChange).toHaveBeenCalledWith(1)
  })

  it('should not increase more than the maxQuantity value', () => {
    const onChange = jest.fn()
    const { getByTestId } = render(
      <Quantity value={1} onChange={onChange} max={1} />
    )

    const increaseButton = getByTestId('increase-quantity')
    increaseButton.click()

    expect(onChange).toHaveBeenCalledWith(1)
  })

  it('should not decrease to less than one', () => {
    const onChange = jest.fn()
    const { getByTestId } = render(
      <Quantity value={1} onChange={onChange} max={100} />
    )

    const decreaseButton = getByTestId('decrease-quantity')
    decreaseButton.click()

    expect(onChange).toHaveBeenCalledWith(1)
  })
})
