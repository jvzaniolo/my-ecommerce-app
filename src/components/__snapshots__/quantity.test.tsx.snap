// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantity should render the quantity with initial props 1`] = `
<div>
  <div
    class="chakra-stack css-84zodg"
  >
    <button
      aria-disabled="true"
      aria-label="Decrease"
      class="chakra-button css-p7mkev"
      data-testid="decrease-quantity"
      disabled=""
      role="button"
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-13otjrl"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 13H5v-2h14v2z"
        />
      </svg>
    </button>
    <input
      aria-valuemax="100"
      aria-valuemin="1"
      aria-valuenow="1"
      aria-valuetext="1"
      autocomplete="off"
      autocorrect="off"
      class="chakra-input css-0"
      inputmode="decimal"
      pattern="[0-9]*(.[0-9]+)?"
      role="spinbutton"
      type="text"
      value="1"
    />
    <button
      aria-label="Increase"
      class="chakra-button css-p7mkev"
      data-testid="increase-quantity"
      role="button"
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-13otjrl"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </button>
  </div>
</div>
`;
